replica/table.cc

```diff
@@ -2837,4 +2838,210 @@ bool table::requires_cleanup(const sstables::sstable_set& set) const {
+
+        if (next_group < _current_group) {
+            on_internal_error(tlogger, format("Token {} caused incremental selection of compaction groups to go backwards, " \
+                                       "the range of current group is {}", t, _cg_sstable_sets[_current_group].first));
+        }
+
+        if (next_group > _current_group) {
+            _current_selector = make_selector(next_group);
+            _current_group = next_group;
+        }
+
+        return *_current_selector;
+    }
+
+    const dht::token& first_token_of_next_group(const dht::token& t) {
+        unsigned group = group_of(_cg_sstable_sets, t);
+
+        if (group == _cg_sstable_sets.size() - 1) {
+            return dht::maximum_token();
+        }
+        return _cg_sstable_sets[group + 1].first.start()->value();
+    }
+public:
+    incremental_selector(const schema& schema, const sstable_sets_t& sets)
+            : _schema(schema)
+            , _cg_sstable_sets(sets)
+            , _current_group(group_of(sets, dht::minimum_token()))
+            , _current_selector(make_selector(_current_group)) {
+    }
+
+    incremental_selector(const incremental_selector&) = delete;
+    incremental_selector(incremental_selector&&) = delete;
+
+    virtual std::tuple<dht::partition_range, std::vector<shared_sstable>, dht::ring_position_ext> select(const dht::ring_position_view& pos) override {
+        const dht::token& t = pos.token();
+        auto& selector = maybe_advance_selector(t);
+
+        auto ret = selector.select(pos);
+
+        auto cmp = dht::ring_position_comparator(_schema);
+
+        // Next position must always be the minimum of next_position returned by incremental selector *and*
+        //      first position of next compaction group.
+        // That guarantees one group's selector won't prevent incremental selection to continue on the
+        // subsequent groups once the current one is exhausted.
+        dht::ring_position_view next_position = ret.next_position;
+        dht::ring_position_view first_pos_of_next_group = dht::ring_position_view(first_token_of_next_group(t));
+        if (cmp(first_pos_of_next_group, next_position) < 0) {
+            next_position = first_pos_of_next_group;
+        }
+
+        // By restricting current range to current position, incremental_selector will refresh the set of
+        // SSTables *and* next_position on every new token.
+        // That's important as SSTable's token range may cross the token range boundary of the compaction
+        // group(s) it's assigned to, e.g. after a split.
+        // next position is still used by the combined reader, to avoid calling the selector needlessly.
+        const dht::partition_range current_range = dht::partition_range::make_singular(dht::ring_position::starting_at(t));
+
+        return std::make_tuple(std::move(current_range), ret.sstables, dht::ring_position_ext(next_position));
+    }
+};
```

Comment:
We should limit the range scanned from each sstable to only the owned range, to prevent unclean compaction groups from supplying bad data. It's not a problem for vnodes, because the coordinator will not ask for such ranges, but here we have ranges that are owned by the node, not owned by the compaction group, but present in its sstables. We may need to extend the incremental_selector interface to return a range along with each sstable.